Add format-icons flag for nerdfonts instead of unicode

afbjorklund opened the pull request:
https://www.nerdfonts.com/

Icons from the VS Code glyphs:

* https://microsoft.github.io/vscode-codicons/dist/codicon.html

Or from the GitHub octicons:

* https://primer.style/design/foundations/icons (from "[octocat](https://octodex.github.com/)")

Screenshots with [Ubuntu Mono](https://www.programmingfonts.org/#ubuntu):

* unicode

![gotestsum-unicode](https://github.com/gotestyourself/gotestsum/assets/10364051/663c9515-f570-4dc4-b929-9a80fe7e9667)

* nerdfonts

![gotestsum-nerdfonts](https://github.com/gotestyourself/gotestsum/assets/10364051/2a36e960-5393-488e-89d2-d532a8061d8f)

Other screenshots with [Roboto Mono](https://www.programmingfonts.org/#roboto).

dnephin reviewed (approved):
Thank you for the PR! I haven't installed the fonts yet, but the screenshot looks good.

There were some conflicts due to changes from #359. From the discussion on that PR I thought it would be good to add an alternative icon that worked better with screen readers.  Now that we have so many options for icons I thought it would be better to select the icons using a `--format-icons` flag instead of separate boolean flags for each.

I pushed those changes to this PR as part of resolving the conflicts. I should have tried to keep more of those changes for a separate commit after resolve. I'll do that next time. 

How does this look to you?

afbjorklund commented:
Having shared parameter for icon formats sounds good to me

afbjorklund commented:
Funny story: when I went to school, the teacher gave me either 𝓡 (right) or ✓ (wrong).

afbjorklund commented:
The format from "testname", while not strictly icons, also has its merits...

![image](https://github.com/gotestyourself/gotestsum/assets/10364051/34651f20-1517-4720-9c94-4f511c2a821a)

```go
		return icons{
			pass: color.GreenString("PASS"),
			skip: color.YellowString("SKIP"),
			fail: color.RedString("FAIL"),
		}.forAction
```

afbjorklund commented:
And these icons are rather subtle, for higher resolution screens:

![gotestsum-emoticons](https://github.com/gotestyourself/gotestsum/assets/10364051/39a62e26-24f6-46f6-935e-5f6add16ebf6)

afbjorklund commented:
Now up to 10 different icons, if you count the emojis (hivis)

* default
* codicons
* codicons-circle
* octicons
* octicons-circle
* octicons-filled
* emoticons-circle
* emoticons-filled
* text

<details>
<img src="https://github.com/gotestyourself/gotestsum/assets/10364051/72eeb9f6-e0a0-429b-bf32-1018026031c7">
</details>

afbjorklund commented:
Rebased and added proper help and environment variable for the format icons:

![gotestsum-formaticons](https://github.com/gotestyourself/gotestsum/assets/10364051/f1a4161b-e1a4-4af2-9f6f-fcfe530c3325)

afbjorklund commented:
Side note: you can check the result of the "text" icons with a screen reader:

```
gotestsum --format-icons text  --hide-summary all | espeak -k20
```

The `-k20` causes it to indicate the TEST capital letters with using an increased pitch...

You can also add `-x -q`, if you just want to see and not hear the phoneme mnemonics

-----

```
sk'Ip_
sk'Ip_ s,i:;,Emd'i: sl'aS Int'3:n@L sl'aS s'Igna#lh,and@L@dr,aIv3
p'aas_ s,i:;,Emd'i:_:_: k'aSt
p'aas_ s,i:;,Emd'i: sl'aS t'u:l sl'aS m'eItrIks_:_: k'aSt
sk'Ip_ k'0ntrIb sl'aS n'oUtI2f,aI
p'aas_ s,i:;,Emd'i: sl'aS t'u:l sl'aS sl'oUI2st_:_: k'aSt
sk'Ip_ Int'3:n@L sl'aS d'0traIt3
sk'Ip_ Int'3:n@L sl'aS t'Ekst
sk'Ip_ Int'3:n@L sl'aS l'0g
p'aas_ Int'3:n@L sl'aS 'agrI2g,eIt_:_: k'aSt
p'aas_ Int'3:n@L sl'aS dZ'u:nItksm@L_:_: k'aSt
p'aas_ Int'3:n@L sl'aS f'aIlw0tS3_:_: k'aSt
p'aas_ t'EstdZs@n_:_: k'aSt

d'Vn_ w'0nh'VndrI2d@n n'aInti; 'eIt t'Ests
t'u: sk'Ipt In z'i@roU pOInt f'o@ n'aIn t'u: z
```

```
∅  .
∅  cmd/internal/signalhandlerdriver
✓  cmd (cached)
✓  cmd/tool/matrix (cached)
∅  contrib/notify
✓  cmd/tool/slowest (cached)
∅  internal/dotwriter
∅  internal/log
∅  internal/text
✓  internal/junitxml (cached)
✓  internal/aggregate (cached)
✓  internal/filewatcher (cached)
✓  testjson (cached)

DONE 198 tests, 2 skipped in 0.460s
```

afbjorklund commented:
Final tweak, added alternative.

```
Format icons:
    default                  the original unicode (✓, ∅, ✖)
    alternative              an alternative unicode (✓, ‐, ✗)
    hivis                    higher visibility unicode (✅, ➖, ❌)
    text                     simple text characters (PASS, SKIP, FAIL)
...
```

dnephin commented on cmd/main.go:
```diff
@@ -145,6 +149,20 @@ Formats:
     standard-verbose         standard go test -v format
 
+Format icons:
+    default                  the original unicode (✓, ∅, ✖)
+    alternative              an alternative unicode (✓, ‐, ✗)
+    hivis                    higher visibility unicode (✅, ➖, ❌)
+    text                     simple text characters (PASS, SKIP, FAIL)
+  * nerdfonts                requires a font from https://www.nerdfonts.com/
+    codicons                   
+    codicons-circle            
+    octicons                   
+    octicons-circle            
+    octicons-filled            
+    emoticons-circle         󰇵 󰇶 󰇸
+    emoticons-filled         󰱱 󰱴 󰱶
```
Adding `text` sounds great! 

I'm concerned now that maybe we have too many options, at least for the initial release. It can be a lot to choose from, and some of them don't look all that different. `default`, `hivis`, `text` are somewhat distinct from each other.

`alternative` isn't much different from `default` and `hivis`, maybe we could leave that one out for now?

For the `nerdfonts`, maybe we could pick just one of each of the `codicons`, `octicons` and `emoticons` variants? That would give us 6 options in total. We can always add more later.

If you had to pick between the default, `-circle`, and `-filled` variants, which do you think we should take? Maybe  `codicons`, `octicons-circle`, and `emoticons-filled` ?

afbjorklund replied:
Okay, will pick out some favorites. Possibly rename the finalists.

afbjorklund replied:
![gotestsum-icons](https://github.com/gotestyourself/gotestsum/assets/10364051/8ce0f8a3-6c09-4f3f-8789-3ae4cf48a52f)

I like the codicons because those are the ones I started the PR with, the octicons because they are consistent between the "circle"/outline and the "filled", and the emoticons because they are fun/colorful and whimsical.
The ones that I left out are either too similar to the default ones and not worth the hassle, or they have poor usability and visibility since the color is too vague or the form is not distinct. Added unicode from go-test-report.

Will use some better names, without dashes.

afbjorklund replied:
The idea was that the icons would be set in environment (or configuration?), and wouldn't be in the way.

afbjorklund replied:
Reduced it down to 4, can add text and emoticons later.

afbjorklund replied:
Might want to add some kind of user configuration to gotestsum, so that they can set their own icons.

dnephin reviewed (commented):
Thanks again! I was away for a bit, so just getting back to this now.

dnephin reviewed (approved):
Awesome!
